reprclust/reproducibility.py

```diff
@@ -0,0 +1,212 @@
+"""
+Module to compute stability estimates of clustering solutions using a method
+inspired by Lange et al. (2004), and Yeo et al. (2011). The procedure
+cross-validates across subjects in the following way: the datasets are divided
+into training and test set; clustering is performed on the training set, and
+the solution is predicted on the test set. Then, the clustering solution is
+computed on the test set, and this is compared to the predicted one using the
+Adjusted Rand Index, the Adjusted Mutual Information, and the Instability Score
+(Lange et. al, 2004) as metrics. Clustering solutions are swept from k=2 to a
+maximum k defined by the user.
+
+At the moment the following clustering algorithms are implemented:
+    - k-means
+    - Gaussian Mixture Models
+    - Ward Clustering (structured and unstructured)
+    - complete linkage with correlation distance
+
+References:
+-----------
+Lange, T., Roth, V., Braun, M. and Buhmann J. (2004)
+"Stability-based validation of clustering solutions."
+Neural computation 16, no. 6 (2004): 1299-1323.
+
+Thomas Yeo, B. T., Krienen, F. M., Sepulcre, J., Sabuncu, M. R., Lashkari, D.,
+Hollinshead, M., et al. (2011).
+"The organization of the human cerebral cortex estimated by intrinsic
+functional connectivity."
+Journal of Neurophysiology, 106(3), 1125–1165. doi:10.1152/jn.00338.2011
+"""
+import copy
+
+from joblib import Parallel, delayed
+
+import numpy as np
+
+from reprclust.cluster_methods import ClusterMethod
+from reprclust.cluster_metrics import ari, ami
+
+# this must be outside the class to allow parallelization
+def _run_fold(self, train, test):
+    """Run reproducibility algorithm on one fold for all the ks"""
+    # initialize methods
+    cm_train = self._cluster_method
+    # copy the method for the testing part too
+    # XXX: am I killing myself with all these copies?
+    cm_test = copy.deepcopy(cm_train)
+
+    # XXX: this should change depending on the type of data
+    # link clustering method to data
+    data_train = [self._data[tr_idx] for tr_idx in train]
+    data_test = [self._data[te_idx] for te_idx in test]
+
+    if self._stack:
+        data_train = np.vstack(data_train)
+        data_test = np.vstack(data_train)
+    else:
+        data_train = np.mean(np.dstack(data_train), axis=-1)
+        data_test = np.mean(np.dstack(data_test), axis=-1)
+
+    cm_train(data_train)
+    cm_test(data_test)
+
+    # allocate storing dictionary
+    result_fold = {}
+    for metric in self._cluster_metrics:
+        result_fold[metric.__name__] = \
+            np.vstack((self._ks, np.zeros(len(self._ks))))
+        if self._ground_truth is not None:
+            result_fold[metric.__name__ + '_gt'] = \
+                np.vstack((self._ks, np.zeros(len(self._ks))))
+
+    # Step 1. Clustering on training/test set and prediction
+    for k in self._ks:
+        # cluster on training set
+        cm_train.cluster(k)
```

Comment:
add option `fully_train` for cases like hierarchical clustering, which will be ignored by other methods